### PR TITLE
Change project-cone-sync to not dump projects outside of the cone.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The prior way we did project-cone syncing was to only include the checksums for the project (and dependents) in the total checksum we sent over.  This had the problem that on the OOP side all other projects would get dumped (including all the project trackers and whatnot) losing all that information we had.  If another OOP request was made for the full solution, we'd have to go generate those again.  

The new approach here works subtly different.  When syncing a project-cone, we ensure that we only get the up to date checksum for that project (and it's dependents).  For any other projects, instead of dropping them, we include the most recently computed checksum for it so that we keep what we have on the OOP side once sync'ed over.

### CODE REVIEW

**Review by ryzngard — APPROVED** [review]
I can't comment on if this is the direction we want to go, but can comment that the code does what is described 👍 

Worth another set of eyes

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ready for review.

**Comment by allisonchou** [comment]
Re-targeting from `release/dev17.3` to `main-vs-deps` so we can delete the release/dev17.3 branch.

**Comment by CyrusNajmabadi** [comment]
Ping @jasonmalinowski

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs [hunk 1]
```diff
@@ -119,14 +119,29 @@ public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationTo
...
-                                                                .ToArray();
+                    var projectChecksumTasks = orderedProjectIds
+                        .Select(id => (state: ProjectStates[id], mustCompute: projectsToInclude == null || projectsToInclude.Contains(id)))
+                        .Where(t => RemoteSupportedLanguages.IsSupported(t.state.Language))
+                        .Select(async t =>
+                        {
+                            // if it's a project that's specifically in the sync'ed cone, include this checksum so that
+                            // this project definitely syncs over.
+                            if (t.mustCompute)
+                                return await t.state.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs:133]: Small allocation nit: you could remove the await here and make the other returns be a Task.FromResult.

File: src/VisualStudio/Core/Test.Next/Services/SolutionServiceTests.cs [hunk 2]
```diff
@@ -628,22 +637,28 @@ public async Task TestPartialProjectSync_GetDependentProjects2()
...
+            await solution.AppendAssetMapAsync(map, project2.Id, CancellationToken.None);
             var project2Checksum = await solution.State.GetChecksumAsync(project2.Id, CancellationToken.None);
-            var project2SyncedSolution = await remoteWorkspace.GetSolutionAsync(assetProvider, project2Checksum, fromPrimaryBranch: false, workspaceVersion: -1, project2.Id, CancellationToken.None);
-            Assert.Equal(1, project2SyncedSolution.Projects.Count());
-            Assert.Equal(project2.Name, project2SyncedSolution.Projects.Single().Name);
+            var project2SyncedSolution = await remoteWorkspace.GetSolutionAsync(assetProvider, project2Checksum, fromPrimaryBranch: false, workspaceVersion: -1, CancellationToken.None);
+            Assert.Equal(2, project2SyncedSolution.Projects.Count());
+            AssertEx.SetEqual(new[] { project2.Name, project3.Name }, project2SyncedSolution.Projects.Select(p => p.Name));
 
+            // if we then sync just P1, we should have 3 projects synved over now.
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test.Next/Services/SolutionServiceTests.cs:656]:
```suggestion
            // if we then sync just P1, we should have 3 projects synced over now.
```

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**